### ISSUE TITLE
Parallel processing with the py script

### ISSUE THREAD

**Issue by rbutleriii** [yunguan-wang/Spacia#52]
Hi @yunguan-wang, 

I noticed in the manuscript it noted that parallel processing actually resulted in slower jobs (presumably implemented in R), but spacia.py has a worker pool. It appears to batch mcmc models?

I have a cluster, so I can generally use multiple jobs for situations where there is 1 sender gene and 1 receiver gene, but is it then faster to run more threads in cluster mode, or pca mode? In this case does the memory scale linearly with the number of cores?

**Comment by yunguan-wang** [comment]
Yes you can use one job per 1 receiver gene, but running spacia on 1 sender gene is not recommended. This is because, even though gene effect beta is independently calculated for each sender gene, the cell proximity effect, b is calculated considering all sender genes in the job.
Spacia do not use much memory, and thus memory should not be an issue during scaling up.

**Comment by rbutleriii** [comment]
Thanks @yunguan-wang, I do seem to be encountering a memory issue. I get an error:
```sh
slurmstepd: error: poll(): Bad address
```
which then hangs for several days until it times out. Checking the allocations for the job:
```sh
sacct -u rrbutler -j 46800990 -o JOBID,JobName,MaxRSS,nCPUs,CPUTime
JobID           JobName     MaxRSS      NCPUS    CPUTime
------------ ---------- ---------- ---------- ----------
46800990     Apoe_L4_I+                     2 10-08:28:04
46800990.ba+      batch  33554260K          2 10-08:28:06
46800990.ex+     extern          0          2 10-08:28:04
```
It does appear to max at 32GB. In this case I am running it with `--mem=32G` and `-c 1`. This dataset has been filtered to only have these cell types for comparison in the meta and counts files.       
```r
ASC L4 IT CTX
3426       450
```
This is running it with the `-sc 'L4 IT CTX'` `-rc 'ASC'` `-sf Apoe` settings with no receiver specified. If I drop the `-sf Apoe` for a general clustering-driven analysis, then it runs out of memory with `--mem=96GB` after a few days. Another consideration is that each experimental group has three slides of CosMx data that is concatenated together with spacing in the x and y positions, but I think the distance cutoff is well below the spacing between the images. It ran successfully for 3 of 4 experimental groups, just this one is taking much longer than the others.

In terms of running Spacia on a single sender gene, I have a liana tensor-based analysis in a matched snRNAseq dataset that implicates `Apoe -> Abca1` as changing between wildtype and disease models. The goal would be to validate that effect in a spatial CosMx set looking at `-sf Apoe` and `-rf Abca1`. Unfortunately the `Abca1` gene is not present in the CosMx probeset, so instead I am running it without the `rf` to see if there are any significant modules associated with `Apoe` as a sender. It looks like there are so the goal would then be to run Spacia again for each receiver gene from the significant receiver module with Apoe pairwise, so I can plot all of their beta distributions similar to Fig 6e. The distribution of `Primary_instance_score` like in Fig 6a would only be done looking at the cell types without any sender or receiver features specified. Does that make sense?

**Comment by rbutleriii** [comment]
I think I may have figured it out here, looking at the logs, I noticed that for certain groups (`WT-C31` and `WT-V`) the `L4 IT CTX` --> `ASC` direction automatically determined a maximal distance that was much, much higher, and those were the ones with the multi day run times:
```sh
grep 'Maximal distance' */*ASC*/*spacia_log.txt
ASC-2-L4_IT_CTX/Efna5_ASC-2-L4_IT_CTX-2.PS19-C31/spacia_log.txt:16:18:24,INFO:::Maximal distance for 10 expected neighbors is 1242.25
ASC-2-L4_IT_CTX/Efna5_ASC-2-L4_IT_CTX-2.PS19-V/spacia_log.txt:16:18:25,INFO:::Maximal distance for 10 expected neighbors is 1249.40
ASC-2-L4_IT_CTX/Efna5_ASC-2-L4_IT_CTX-2.WT-C31/spacia_log.txt:16:18:26,INFO:::Maximal distance for 10 expected neighbors is 1072.18
ASC-2-L4_IT_CTX/Efna5_ASC-2-L4_IT_CTX-2.WT-V/spacia_log.txt:16:18:26,INFO:::Maximal distance for 10 expected neighbors is 1134.22
ASC-2-L4_IT_CTX/general_ASC-2-L4_IT_CTX.PS19-C31/spacia_log.txt:16:18:20,INFO:::Maximal distance for 10 expected neighbors is 1242.25
ASC-2-L4_IT_CTX/general_ASC-2-L4_IT_CTX.PS19-V/spacia_log.txt:16:18:20,INFO:::Maximal distance for 10 expected neighbors is 1249.40
ASC-2-L4_IT_CTX/general_ASC-2-L4_IT_CTX.WT-C31/spacia_log.txt:16:18:26,INFO:::Maximal distance for 10 expected neighbors is 1072.18
ASC-2-L4_IT_CTX/general_ASC-2-L4_IT_CTX.WT-V/spacia_log.txt:16:18:25,INFO:::Maximal distance for 10 expected neighbors is 1134.22
L4_IT_CTX-2-ASC/Apoe_L4_IT_CTX-2-ASC.PS19-C31/spacia_log.txt:16:18:19,INFO:::Maximal distance for 10 expected neighbors is 1261.89
L4_IT_CTX-2-ASC/Apoe_L4_IT_CTX-2-ASC.PS19-V/spacia_log.txt:16:18:19,INFO:::Maximal distance for 10 expected neighbors is 1257.63
L4_IT_CTX-2-ASC/Apoe_L4_IT_CTX-2-ASC.WT-C31/spacia_log.txt:22:39:41,INFO:::Maximal distance for 10 expected neighbors is 137220.51
L4_IT_CTX-2-ASC/Apoe_L4_IT_CTX-2-ASC.WT-V/spacia_log.txt:16:18:19,INFO:::Maximal distance for 10 expected neighbors is 14762.13
L4_IT_CTX-2-ASC/general_L4_IT_CTX-2-ASC.PS19-C31/spacia_log.txt:16:18:19,INFO:::Maximal distance for 10 expected neighbors is 1261.89
L4_IT_CTX-2-ASC/general_L4_IT_CTX-2-ASC.PS19-V/spacia_log.txt:16:18:19,INFO:::Maximal distance for 10 expected neighbors is 1257.63
L4_IT_CTX-2-ASC/general_L4_IT_CTX-2-ASC.WT-C31/spacia_log.txt:22:39:40,INFO:::Maximal distance for 10 expected neighbors is 137220.51
L4_IT_CTX-2-ASC/general_L4_IT_CTX-2-ASC.WT-V/spacia_log.txt:12:28:24,INFO:::Maximal distance for 10 expected neighbors is 14762.13
``` 
As noted above I concatenated slide images, and I think even though they were very spaced apart, there was one group where the calculated distance spanned the whole 100k+ pixel distance! 

I will rerun the two with odd distances set at `--dist_cutoff 1250` pixels, since that seems in line with the other two experimental groups (`PS19-V` and `PS19-C31`) and see if that fixes the memory/run time. (Edit: Yep, 7 hour run times now).

1. This does bring up a question though. When trying to compare across experimental groups, should they all be set to the same distance? In this case, astrocytes are mobile cells, so the spatial proximity of the two cell types may be biologically relevant in wildtype and disease mice. I could also standardize by `--n_neighbors`  but I imagine that may result in some pairings adopting a range of (potentially unnecessary) larger distances like above.

2. Another question more in line with parallelism, is what to do with slide replicates. I have three slides per group, but if I run them separately I am not sure how to combine the b, beta and p values effectively. Concatenating them does work, but runs the risk of grabbing neighbors from another slide for receivers along the edge, unless I manually set the max distance (say 1250) and space the slides further than that (say 5000 pixels). Thoughts?

**Comment by yunguan-wang** [comment]
@rbutleriii 
Thanks for the interesting diagnosis. I am happy to see that you have found the issue!
For your questions
1. This is more on the biology side. If you are comparing CCC between the same cell types in different experiments, I would keep the distance as a constant. You can specify it in the `spacia.py` script.
2. @wtwt5237 @wooyongc This is a very good question. In the spacia paper, we combine Spacia results based on individually filtering on b values first (pvalue of b <= cutoff). Biologically this is roughly equivalent to filtering on **receiving genes** identified in each individual replicate. Then you can do filtering or averaging on the beta values, for each individual interactions.

**Comment by rbutleriii** [comment]
The answer to 1. and setting the `--dist_cutoff` did resolve the parallelism question, so we could close this issue and break out the multi-slide concept to a new issue if you like.
